File: Python/search/subsets_ii.py
from typing import *

def subsetsWithDup(nums: List[int]) -> List[List[int]]:
    res = list()
    ans = list()
    nums = sorted(nums)
    length = len(nums)
    def dfs(choose_pre, point):
        if point == length:
            res.append(ans.copy())
            return
        dfs(False, point + 1)
        if not choose_pre and point != 0 and nums[point] == nums[point - 1]:
            return
        ans.append(nums[point])
        dfs(True, point + 1)
        del ans[-1]
    
    dfs(False, 0)
    return res

File: Java/src/com/dbc/algorithm/Search/SubsetsII.java
package com.dbc.algorithm.Search;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SubsetsII {
    private List<List<Integer>> res = new ArrayList<>();
    private List<Integer> ans = new ArrayList<>();
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        dfs(false, nums, 0);
        return res;
    }

    private void dfs(boolean choosePre, int[] nums, int point) {
        if (nums.length == point) {
            res.add(new ArrayList<>(ans));
            return;
        }
        dfs(false, nums, point + 1);
        if (!choosePre && point != 0 && nums[point] == nums[point - 1]) {
            return;
        }
        ans.add(nums[point]);
        dfs(true, nums, point + 1);
        ans.remove(ans.size() - 1);
    }
}
